### ISSUE TITLE
coop selective element selection mechanism

### ISSUE THREAD

**Issue by DrMeepster** [BEEmod/BEE2-items#3317]
There are many suggestions about coop here that have the same flaw; they take away coop players' ability to choose which part of the puzzle they want to do and their ability to swap places if they get tired. (#2119, #2120, #3266).

I propose a solution where instead of the map maker choosing which player color can interact with an item, they choose which of 2 shapes can interact with an item. Then, at the beginning of a map or at checkpoints, players can use the ping tool to choose a shape.

If a shape is currently picked by no one, it will be gray, and if it is picked by both players it will have orange and blue stripes. Otherwise, it takes its player's color.

If 1 or more shapes are unchosen with both players inside the checkpoint, the door will not open. When there is a player outside of the active checkpoint and a player changes their shape, red outlines will appear around the shapes and testing elements will not update who they affect until both players are in the checkpoint or move on to the next checkpoint.

Testing elements using this system would display their shape on them and the player color they currently affect, preferably while both active and inactive. For example, a selective fizzler affecting Triangle (currently chosen by orange) would have triangles on its field, the field would be orange and its emitter would have orange triangles on it.

My suggestions for what the shapes would actually be are Triangle and Circle.

**Comment by LautaroL20** [comment]
This is a duplicate of: https://github.com/BEEmod/BEE2-items/issues/702, however, i´ve added the description of your issue to the original one. If you have more ideas to share regard to this subject, make sure to comment on issue #702

**Comment by TeamSpen210** [comment]
Not really, this is a seperate thing. This is something I've been keeping in mind when designing items. It's why the seperation checkpoint has buttons, so you can decide which side to exit at. The respawn droppers are also designed so they can handle either player on the sides. This idea would be neat, especially if I built it as a pre-spawn override where you first drop into a selection room, then continue onto the main map.

**Comment by Luke18033** [comment]
This should also affect Aperture Tag (allowing players to choose who has the portal gun and who has the paint gun).

**Comment by Luke18033** [comment]
I thought of a simpler way to implement this, just swap Atlas and P-Body's models if you choose that option at the start of the map. Since internally the players aren't actually swapped, it'd still require modifying every coop-player-detecting item to handle this system and portal colors wouldn't change, but it would mean that visibly we can reuse the existing blue/orange player identification system rather than having to come up with something new.

**Comment by Luke18033** [comment]
Never mind, that would be way too finicky with cosmetics... Maybe we just swap the bots' eye colors around using `prop_dynamic_ornament` overlay models?

**Comment by Luke18033** [comment]
The best way to implement this would probably be to have the test elements physically switch to the corresponding bot's color, we'd need to do that anyway for autoportals. Maybe we should also allow switching mid-level at specific places like checkpoints?

**Comment by vrad-exe** [comment]
I found at one point you can call the `SetTeam` VScript function on co-op players to literally just swap Atlas and P-Body, maybe that would work for this
